### ISSUE TITLE
loading directed graphs as undirected?

### ISSUE THREAD

**Issue by jowens** [gunrock/gunrock#497]
**Summary:**
from @ogreen: "Gunrock seems to be loading directed graphs as undirected graphs and as such is creating inverse edges. This means that we have two results for undirected graphs and none for directed graphs."

@ogreen can you help fill this in? Ask Euna to help?

@sgpyc @neoblizz @porumbes can you help?

**Reproducibility:**
Always or Sometimes

**To Reproduce:**
Detail the steps necessary to reproduce the behavior.

For example:
1. Clone gunrock
2. Switch to the dev-refactor branch

Make sure to:
$ git submodule init
$ git submodule update
etc.
etc.

**Expected Behavior:**

**Actual Results:**
A clear and concise description of what you actually experienced.

**System Configuration:**

GPU: GTX 970
CPU: Intel(R) Xeon(R) CPU E3-1280 v5 @ 3.70GHz
OS: Ubuntu 18.04.1 LTS
Cuda compilation tools, release 10.0, V10.0.130
NVIDIA-SMI 415.13
Driver Version: 415.13
CUDA Version: 10.0
gcc (Ubuntu 6.4.0-17ubuntu1) 6.4.0 20180424

**Comment by eunaeuna** [comment]
This is simply test_bc.cu parameter setting I believe. Currently "--undirected" is set as true.
Is there any particular reason for this?

https://github.com/gunrock/gunrock/blob/master/tests/bc/test_bc.cu, line 799
    // graph construction or generation related parameters
    info -> info["undirected"] = true;  // require undirected input graph

We tested with below fix and it worked fine.
info->info["undirected"] = args -> CheckCmdLineFlag("undirected");

**Comment by neoblizz** [comment]
@sgpyc can you comment on why undirected is forced?
https://github.com/gunrock/gunrock/blob/master/tests/bc/test_bc.cu#L798-L799

**Comment by sgpyc** [comment]
The master branch uses BGL to compute the BC values when src=-1, and BGL uses an algorithm from the paper by Ulrik Brandes, "A Faster Algorithm for Betweenness Centrality", in the Journal of Mathematical Sociology, volume 25 number 2, pages 163--177, 2001. The paper noted "For simplicity, we assume that all graphs are undirected and connected, though they may have loops or multiple edges. Note that our results generalize to directed graphs with only minor modification." 
I think enforcing input graphs to be undirected was to get comparable results with BGL.

In the dev-refactor branch, since we don't use BGL,  the graph is not forced to be undirected any more.

**Comment by jowens** [comment]
@sgpyc is it thus appropriate to use the dev-refactor branch to compute directed-graph BC? should that give us similar performance?

**Comment by neoblizz** [comment]
> The master branch uses BGL to compute the BC values when src=-1, and BGL uses an algorithm from the paper by Ulrik Brandes, "A Faster Algorithm for Betweenness Centrality", in the Journal of Mathematical Sociology, volume 25 number 2, pages 163--177, 2001. The paper noted "For simplicity, we assume that all graphs are undirected and connected, though they may have loops or multiple edges. Note that our results generalize to directed graphs with only minor modification."
> I think enforcing input graphs to be undirected was to get comparable results with BGL.
> 
> In the dev-refactor branch, since we don't use BGL, the graph is not forced to be undirected any more.

@sgpyc Does this mean that we don't HAVE to enforce input graph as undirected? It should still work with directed, right? Or BGL is only for undirected graph?

**Comment by sgpyc** [comment]
Yes, in 'dev-refactor', we don't have to enforce input graph as undirected.
In 'dev' and 'master' branches, BGL is used when `src=-1` (all source BC), so the graph needs to be undirected; if it's not all source BC, I think undirected input is not necessary, but haven't fully tested yet.

**Comment by ogreen** [comment]
You don't need undirected graphs. You simply need to change the direction of the traversal from the children looking to their parents to the parents looking at their children. 
You will also need to look to modify the dependency accumulation to start at one level above the lowest level of leaves. I also don't recommend storing the parents or the children: https://www.researchgate.net/profile/Oded_Green/publication/257719848_Faster_Betweenness_Centrality_Based_on_Data_Structure_Experimentation/links/53f67b9f0cf2fceacc73f5db.pdf
